core/templates/pages/exploration-player-page/services/content-translation-manager.service.ts

```diff
@@ -0,0 +1,212 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service to manage the content translations displayed.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { EventEmitter, Injectable } from '@angular/core';
+
+import { PlayerTranscriptService } from
+  'pages/exploration-player-page/services/player-transcript.service';
+import { StateCard } from
+  'domain/state_card/StateCardObjectFactory';
+import { WrittenTranslation } from
+  'domain/exploration/WrittenTranslationObjectFactory';
+import { SubtitledHtml } from
+  'domain/exploration/SubtitledHtmlObjectFactory';
+import { Schema } from 'services/schema-default-value.service';
+import { SchemaConstants } from
+  'components/forms/schema-based-editors/schema-constants';
+import INTERACTION_SPECS from 'pages/interaction-specs.constants.ajs';
+import { WrittenTranslations } from
+  'domain/exploration/WrittenTranslationsObjectFactory';
+import { SubtitledUnicode } from
+  'domain/exploration/SubtitledUnicodeObjectFactory';
+import { ExtensionTagAssemblerService } from
+  'services/extension-tag-assembler.service';
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ContentTranslationManagerService {
+  _explorationLanguageCode: string;
+  _onStateCardContentUpdateEmitter: EventEmitter<void> = new EventEmitter();
+
+  constructor(
+    private playerTranscriptService: PlayerTranscriptService,
+    private extensionTagAssemblerService: ExtensionTagAssemblerService
+  ) {}
+
+  init(explorationLanguageCode: string) : void {
+    this._explorationLanguageCode = explorationLanguageCode;
+  }
+
+  get onStateCardContentUpdate(): EventEmitter<void> {
+    return this._onStateCardContentUpdateEmitter;
+  }
+
+  displayTranslations(languageCode: string) : void {
+    const cards = this.playerTranscriptService.transcript;
+
+    if (languageCode === this._explorationLanguageCode) {
+      this.playerTranscriptService.restoreImmutably();
+    } else {
+      cards.forEach(
+        card => this._displayTranslationsForCard(card, languageCode));
+    }
+
+    cards.forEach(card => {
+      // Because we do not have a mapping of the learner's input and Oppia's
+      // response to the content from which it originates from, we cannot
+      // easily swap them with translated values. Instead, we opt to clear
+      // the responses.
+      card.getInputResponsePairs().splice(
+        0, card.getInputResponsePairs().length);
+      // Mark card as uncompleted because we clear response pairs.
+      card.markAsNotCompleted();
```

Comment:
Is this correct? The transcript is supposed to be read-only and a representation of the playthrough history, so maybe just keep the existing response pairs but don't display them if the language is changed?

Reply:
Hm, I'm having some trouble implementing this, since the InputResponsePair logic is a little complex and I'm short on time. Any advice on how to proceed? Maybe a compromise is to reload the exploration at that state altogether?

Reply:
I don't think you can "reload the exploration at that state" -- it sounds like the learner can't really go back past that point if you do that? But I see the issue -- we haven't really accounted for this and I agree that it is a bit complex.

Here's a suggestion. How about the following:

- Show the language selector only on the first card of the lesson, but not on other cards.
- Until the learner answers the first question, they can change the language without any issues.
- However, if the learner has progressed beyond the first state (or answers the question in the first state), then navigates back to the first card and changes the language, show a pop-up saying "This will refresh the page and restart the lesson from the beginning. After the page reloads, please change the language to your preferred language. Are you sure?" (It would be better if we could pass the new language in via the URL and auto-set it for them, but if it's too hard then I think this will do for now, though I suggest filing an issue if you're punting this part since the flow here is a bit icky.)

I think this is different from the original plan, but it seems like it will get around this issue of the trickiness of swapping translations in and out. What do you think? Other suggestions welcome.